onnxruntime/contrib_ops/rocm/bert/batched_gemm_softmax_gemm_permute_pipelines.cuh

```diff
@@ -244,6 +283,246 @@ struct GemmSoftmaxGemmPermuteGenericPipeline {
+    constexpr const int kThreadPerBlock = 256;
+    constexpr const int kVecSize = 4;
+
+    auto attn = params->attention;
+    auto [buffer, lengths, strides] = GetRawMaskBufferAddrSizesAndStrides(params->mask_index_buffer, attn);
+    int64_t total_threads = lengths.x * lengths.y * CeilDiv(lengths.z, kVecSize);
+    auto num_blocks = CeilDiv(total_threads, kThreadPerBlock);
+
+    auto mask_filter_value = attn->mask_filter_value;
+    auto cvt = [=] __device__(int v) -> T {
+      return v == 1 ? 0 : mask_filter_value;
+    };
+
+    ConvertToFilledMaskValue<kVecSize><<<num_blocks, kThreadPerBlock, 0, params->Stream()>>>(
+        reinterpret_cast<T*>(params->workspace_buffer),
+        buffer, lengths, strides,  // mask desc
+        cvt);
+
+    return HIP_CALL(hipGetLastError());
+  }
+};
+
+template <typename T, bool USE_BIAS, bool USE_MASK>
+auto GetCKGemmSoftmaxGemmPermuteTypeStringAndOps() {
+  constexpr const int kNumBiasBuffer = static_cast<int>(USE_BIAS) + static_cast<int>(USE_MASK);
+
+  using Nop = ck::tensor_operation::element_wise::PassThrough;
+  using Acc0ElementOp = internal::PreSoftmaxAttentionScoreOp;
+
+  using CKDataType = typename DataTypeAdaptor<T>::type;
+  using D0DataType = typename ck::detail::tuple_concat<
+      std::conditional_t<USE_BIAS, ck::Tuple<CKDataType>, ck::Tuple<>>,
+      std::conditional_t<USE_MASK, ck::Tuple<CKDataType>, ck::Tuple<>>>::type;
+
+  constexpr static auto MaskingSpec =
+      ck::tensor_operation::device::MaskingSpecialization::MaskDisabled;
+
+  std::vector<std::pair<std::string, Op<GemmSoftmaxGemmPermuteParams<T>>>> ret;
+  for (auto&& impl : internal::GetDeviceBatchedGemmSoftmaxGemmPermuteInstances<
+           CKDataType, D0DataType, internal::F32, internal::PreSoftmaxAttentionScoreOp, MaskingSpec>()) {
+    auto type_string = impl->GetTypeString();
+
+    auto invoker = impl->MakeInvokerPointer();
+    auto op = [impl = std::move(impl), invoker = std::move(invoker)](
+                  const GemmSoftmaxGemmPermuteParams<T>* params) -> Status {
+      if constexpr (USE_BIAS) {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(params->bias_buffer == nullptr);
+      } else {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(params->bias_buffer != nullptr);
+      }
+      if constexpr (USE_MASK) {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(
+            !GemmSoftmaxGemmPermuteTunableOp<T>::IsSupportedMaskType(params->attention));
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(params->mask_index_buffer == nullptr);
+      } else {
+        TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(params->mask_index_buffer != nullptr);
+      }
+
+      auto attn = params->attention;
+      int G0 = attn->batch_size;
+      int G1 = attn->num_heads;
+      int M = attn->sequence_length;
+      int N = attn->total_sequence_length;
+      int K = attn->head_size;
+      int O = attn->v_head_size;
```

Comment:
nit: suggest o rename O to V.   O is very similar to 0, it could confuse reader.